libsolidity/analysis/ControlFlowGraph.cpp

```diff
@@ -0,0 +1,552 @@
+	connect(m_currentNode, m_cfg.m_currentFunctionFlow->revert);
+	m_currentNode = newLabel();
+	return false;
+}
+
+bool ControlFlowParser::visit(Block const&)
+{
+	solAssert(!!m_currentNode, "");
+	createLabelHere();
+	return true;
+}
+
+void ControlFlowParser::endVisit(Block const&)
+{
+	solAssert(!!m_currentNode, "");
+	createLabelHere();
+}
+
+bool ControlFlowParser::visit(Return const& _return)
+{
+	solAssert(!!m_currentNode, "");
+	solAssert(!!m_cfg.m_currentFunctionFlow, "");
+	solAssert(!!m_cfg.m_currentFunctionFlow->exit, "");
+	// only the first return statement is interesting
+	if (!m_currentNode->block.returnStatement)
+		m_currentNode->block.returnStatement = &_return;
+	connect(m_currentNode, m_cfg.m_currentFunctionFlow->exit);
+	m_currentNode = newLabel();
+	return true;
+}
+
+
+bool ControlFlowParser::visit(PlaceholderStatement const&)
+{
+	solAssert(!!m_currentNode, "");
+	solAssert(!!m_cfg.m_currentModifierFlow, "");
+
+	connect(m_currentNode, m_cfg.m_currentModifierFlow->placeholderEntry);
+
+	m_currentNode = newLabel();
+
+	connect(m_cfg.m_currentModifierFlow->placeholderExit, m_currentNode);
+	return false;
+}
+
+bool ControlFlowParser::visitNode(ASTNode const& node)
+{
+	solAssert(!!m_currentNode, "");
+	if (auto const* expression = dynamic_cast<Expression const*>(&node))
+		m_currentNode->block.expressions.emplace_back(expression);
+	if (auto const* variableDeclaration = dynamic_cast<VariableDeclaration const*>(&node))
+		m_currentNode->block.variableDeclarations.emplace_back(variableDeclaration);
+	if (auto const* assembly = dynamic_cast<InlineAssembly const*>(&node))
+		m_currentNode->block.inlineAssemblyStatements.emplace_back(assembly);
+
+	return true;
+}
+
+bool ControlFlowParser::visit(FunctionCall const& _functionCall)
+{
+	solAssert(!!m_currentNode, "");
+	if (auto functionType = dynamic_pointer_cast<FunctionType const>(_functionCall.expression().annotation().type))
+		switch (functionType->kind())
+		{
+			case FunctionType::Kind::Revert:
+				solAssert(!!m_cfg.m_currentFunctionFlow, "");
+				solAssert(!!m_cfg.m_currentFunctionFlow->revert, "");
+				ASTNode::listAccept(_functionCall.arguments(), *this);
+				connect(m_currentNode, m_cfg.m_currentFunctionFlow->revert);
+				m_currentNode = newLabel();
+				return false;
+			case FunctionType::Kind::Require:
+			case FunctionType::Kind::Assert:
+			{
+				solAssert(!!m_cfg.m_currentFunctionFlow, "");
+				solAssert(!!m_cfg.m_currentFunctionFlow->revert, "");
+				ASTNode::listAccept(_functionCall.arguments(), *this);
+				connect(m_currentNode, m_cfg.m_currentFunctionFlow->revert);
+				auto nextNode = newLabel();
+				connect(m_currentNode, nextNode);
+				m_currentNode = nextNode;
+				return false;
+			}
+			default:
+				break;
+		}
+	return ASTConstVisitor::visit(_functionCall);
+}
+
+bool CFG::constructFlow(ASTNode const& _astRoot)
+{
+	_astRoot.accept(*this);
+	applyModifiers();
+	return Error::containsOnlyWarnings(m_errorReporter.errors());
```

Comment:
Strictly, this should only return false if it added new errors.

Reply:
I could change that here or I could break on the first error in ``CompilerStack.cpp`` - however, the current behaviour is consistent with the other analysis modules, see e.g. ``PostTypeChecker``, so we should probably either leave it as it is here or change it everywhere...

Reply:
Yeah, it's probably fine.